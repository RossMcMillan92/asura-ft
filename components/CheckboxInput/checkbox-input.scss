@import '../../styles/settings/breakpoints.scss';
@import '../../styles/settings/pallete.scss';

$checkbox-input__size: 30px;
$checkbox-input__size--tablet: 40px;
$checkbox-input__border-width: 2px;
$checkbox-input__tick-size: 50%;

.checkbox-input {
  position: relative;
}

.checkbox-input__input,
.checkbox-input__facade {
  width: $checkbox-input__size;
  height: $checkbox-input__size;

  @media (min-width: $breakpoint-tablet) {
    width: $checkbox-input__size--tablet;
    height: $checkbox-input__size--tablet;
  }
}

.checkbox-input__input {
  display: block;
  opacity: 0;
}

.checkbox-input__facade {
  position: absolute;
  left: 0;
  top: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  pointer-events: none;
  background-color: #fff;
  border-radius: 4px;

  // Animated fill
  &:before {
    content: '';
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background-color: $base-theme-color;
    border-radius: 3px;
    transform: scale(0);
    transition: transform 0.25s ease-out 0.2s;

    .checkbox-input.is-checked & {
      transition: transform 0.25s ease-in;
      transform: scale(1);
    }

    .checkbox-input__input:hover ~ &,
    .checkbox-input__input:focus ~ & {
      background-color: darken($base-theme-color, 10%);
    }
  }

  // Animated border
  &:after {
    content: '';
    position: absolute;
    top: $checkbox-input__border-width;
    left: $checkbox-input__border-width;
    width: calc(100% - #{($checkbox-input__border-width * 2)});
    height: calc(100% - #{($checkbox-input__border-width * 2)});
    border-radius: 2px;
    transform: scale(1);
    box-shadow: 0 0 0 $checkbox-input__border-width #afafaf;
    transition: all 0.25s ease-in, border-color 0s ease;

    .checkbox-input__input:focus ~ &,
    .checkbox-input.is-checked &,
    .checkbox-input:hover &,
    .checkbox-input:focus-within & {
      box-shadow: 0 0 0 $checkbox-input__border-width $base-theme-color;
    }

    .checkbox-input.is-checked & {
      transition: all 0.5s ease-out 0.2s, border-color 0s ease;
      transform: scale(1.75);
      opacity: 0;
    }
  }
}

.checkbox-input__tick {
  position: relative;
  width: $checkbox-input__tick-size;
  height: $checkbox-input__tick-size;
  z-index: 3;
  transform: scale(0);
  transition: transform 0.25s ease-in;

  .checkbox-input.is-checked & {
    transition: transform 0.25s ease-out 0.25s;
    transform: scale(1);
  }

  .checkbox-input.is-checked
    .checkbox-input__input:focus
    ~ .checkbox-input__facade
    &,
  .checkbox-input.is-checked
    .checkbox-input__input:hover
    ~ .checkbox-input__facade
    & {
    transition: transform 0.25s ease-out 0.25s;
    transform: scale(1.2) rotate(5deg);
  }
}
