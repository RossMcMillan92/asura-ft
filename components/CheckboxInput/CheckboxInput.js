import './checkbox-input.scss'

import PropTypes from 'prop-types'
import React from 'react'
import classnames from 'classnames'

const CheckboxInput = ({ checked, className, id, name, onChange, value }) => (
  <div
    className={classnames(
      'checkbox-input',
      { 'is-checked': checked },
      className
    )}
  >
    <input
      className={'checkbox-input__input'}
      id={id}
      type="checkbox"
      name={name}
      value={value}
      checked={!!checked}
      onChange={onChange}
    />
    <div className={'checkbox-input__facade'}>
      <CheckboxTickIcon />
    </div>
  </div>
)

const CheckboxTickIcon = () => (
  <svg
    className="checkbox-input__tick"
    viewBox="0 0 1024 1024"
    width="1024"
    height="1024"
  >
    <path
      className="path0"
      mi-d="M864 128l-480 480-224-224-160 160 384 384 640-640z"
      fill="#fff"
      d="M864 128l-480 480-224-224-160 160 384 384 640-640z"
    ></path>
  </svg>
)

CheckboxInput.propTypes = {
  checked: PropTypes.bool.isRequired,
  className: PropTypes.string,
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.string.isRequired
}

export default CheckboxInput
