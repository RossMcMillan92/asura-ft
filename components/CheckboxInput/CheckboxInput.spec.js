import { mount, shallow } from 'enzyme'

import CheckboxInput from './CheckboxInput'
import React from 'react'
import { act } from 'react-dom/test-utils'

const props = {
  checked: false,
  id: 'checkbox-id',
  name: 'checkbox-name',
  onChange: jest.fn(),
  value: 'value'
}
const getWrapper = additionalProps =>
  mount(<CheckboxInput {...props} {...additionalProps} />)
const getShallowWrapper = additionalProps =>
  shallow(<CheckboxInput {...props} {...additionalProps} />)

describe('CheckboxInput', () => {
  it('matches the snapshot', () => {
    const wrapper = getShallowWrapper()
    expect(wrapper).toMatchSnapshot()
  })

  describe('when changed', () => {
    it('calls the passed onChange prop', () => {
      const wrapper = getWrapper()
      act(() => {
        wrapper.find('input').simulate('change')
      })
      expect(props.onChange).toBeCalled()
    })
  })

  describe('when the prop checked is passed as true', () => {
    const checked = true
    const wrapper = getWrapper({ checked })

    it('sets the input to checked', () => {
      expect(wrapper.find('input').prop('checked')).toBe(checked)
    })

    it('appends the is-checked class to the className', () => {
      expect(wrapper.find('.checkbox-input').prop('className')).toContain(
        `checkbox-input is-checked`
      )
    })
  })

  describe('when a className prop is passed', () => {
    it('appends the class to the className', () => {
      const className = 'additional-styling'
      const wrapper = getWrapper({ className })
      expect(wrapper.find('.checkbox-input').prop('className')).toContain(
        `checkbox-input ${className}`
      )
    })
  })
})
