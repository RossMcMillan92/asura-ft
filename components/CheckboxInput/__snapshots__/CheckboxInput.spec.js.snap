// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxInput matches the snapshot 1`] = `
<div
  className="checkbox-input"
>
  <input
    checked={false}
    className="checkbox-input__input"
    id="checkbox-id"
    name="checkbox-name"
    onChange={[MockFunction]}
    type="checkbox"
    value="value"
  />
  <div
    className="checkbox-input__facade"
  >
    <CheckboxTickIcon />
  </div>
</div>
`;
